thunder/transforms/quantization.py

```diff
@@ -294,3 +295,214 @@ def transform_traces_pre_prologue(self, prologue_trace, computation_trace, epilo
+        self.lora_dropout = lora_dropout
+        self.lora_linear_names = set()
+        self.lora_linear_states = {}
+
+    def transform_module(self, model: thunder.ThunderModule):
+        self.thunder_module = model
+        shared_names = model._get_shared_names()
+        processed_names = set()
+
+        def convert_linear_submodule(tm, name, r):
+            self.lora_linear_names.add(name)
+            weight_name = f"{name}.weight"
+            processed_copies = shared_names[weight_name] & processed_names
+            if processed_copies:
+                copy_name = next(iter(processed_copies))
+                self.lora_linear_states[weight_name] = self.lora_linear_states[copy_name]
+                tm._overrides_parameters[weight_name] = tm._overrides_parameters[copy_name]
+                tm._overrides_parameters[f"{weight_name}.loraA"] = tm._overrides_parameters[f"{copy_name}.loraA"]
+                tm._overrides_parameters[f"{weight_name}.loraB"] = tm._overrides_parameters[f"{copy_name}.loraB"]
+                processed_names.add(weight_name)
+                return
+
+            w = tm.get_parameter(weight_name)
+            in_features, out_features = w.shape[0], w.shape[1]
+
+            loraA = torch.nn.Parameter(torch.empty((r, in_features)))
+            torch.nn.init.kaiming_uniform_(loraA, a=math.sqrt(5))
+            loraB = torch.nn.Parameter(torch.empty((out_features, r)))
+            torch.nn.init.zeros_(loraB)
+
+            tm._overrides_parameters[weight_name] = w
+            tm._overrides_parameters[f"{weight_name}.loraA"] = loraA
+            tm._overrides_parameters[f"{weight_name}.loraB"] = loraB
+
+            self.lora_linear_states[weight_name] = {
+                "dtype": w.dtype,
+                "shape": tuple(w.shape),
+                "loraA.dtype": loraA.dtype,
+                "loraA.shape": loraA.shape,
+                "loraB.dtype": loraB.dtype,
+                "loraB.shape": loraB.shape,
+                "device": getattr(w, "_thunder_device", w.device),
+            }
+            processed_names.add(name)
+
+        for n, submodule in model._model.named_modules():
+            if isinstance(submodule, torch.nn.Linear):
+                convert_linear_submodule(model, n, self.r)
+
+    def transform_state_dict_for_submodule(
+        self, model: thunder.ThunderModule, submodule_name: str, state_dict: dict
+    ) -> dict:
+        if submodule_name not in self.lora_linear_names:
+            return state_dict
+
+        weight_name_full = f"{submodule_name}.weight"
+        w = state_dict["weight"]
+        state_dict = state_dict.copy()
+        state_dict["weight"] = w.to(w.device)
+
+        return state_dict
+
+    def transform_traces_pre_prologue(self, prologue_trace, computation_trace, epilogue_trace, **kwargs):
+        tm = self.thunder_module
+        from thunder.core.trace import tracectx
+
+        checks = get_checks(prologue_trace)
+
+        # do we need this?
```

Comment:
You are right, we don't need this as we are not modifying the tensor metadata (in particular dtype) of w.
We should copy the prologue trace

Reply:
Amazing, I think it is simpler to pass an empty map. Let me do that!